searx/enginelib/__init__.py

```diff
@@ -9,16 +19,158 @@
    - move implementations of the :ref:`searx.engines loader` to a new module in
      the :py:obj:`searx.enginelib` namespace.
 
+-----
+
 """
+from __future__ import annotations
 
+__all__ = ["EngineCache", "Engine", "ENGINES_CACHE"]
 
-from __future__ import annotations
-from typing import List, Callable, TYPE_CHECKING
+from typing import List, Callable, TYPE_CHECKING, Any
+import string
+import typer
+
+from ..cache import ExpireCacheSQLite, ExpireCacheCfg
 
 if TYPE_CHECKING:
     from searx.enginelib import traits
 
 
+ENGINES_CACHE = ExpireCacheSQLite(
+    ExpireCacheCfg(
+        name="ENGINES_CACHE",
+        MAXHOLD_TIME=60 * 60 * 24 * 7,  # 7 days
+        MAINTENANCE_PERIOD=60 * 60,  # 2h
+    )
+)
+"""Global :py:obj:`searx.cache.ExpireCacheSQLite` instance where the cached
+values from all engines are stored.  The `MAXHOLD_TIME` is 7 days and the
+`MAINTENANCE_PERIOD` is set to two hours."""
+
+app = typer.Typer()
+
+
+@app.command()
+def state():
+    """Show state for the caches of the engines."""
+
+    title = "cache tables and key/values"
+    print(title)
+    print("=" * len(title))
+    print(ENGINES_CACHE.state().report())
+    print()
+    title = f"properties of {ENGINES_CACHE.cfg.name}"
+    print(title)
+    print("=" * len(title))
+    print(str(ENGINES_CACHE.properties))
+
+
+@app.command()
+def maintenance(force: bool = True):
+    """Carry out maintenance on cache of the engines."""
+    ENGINES_CACHE.maintenance(force=force)
+
+
+class EngineCache:
```

Comment:
> Implements a key/value store in class searx.cache.ExpireCacheSQLite. The key/value pairs are automatically deleted after a specified period of time (expire) .. similar to the expire already known from redis or valkey.
>
As far as I know, SQLite is not really intended to be used as a key value store. I think it's okay to have it as a fallback, but we should always use Redis if it's available because its performance will be much better as it's actually just made for that single purpose. So I think this should be similar to what I started to implement in https://github.com/searxng/searxng/pull/4314 with different engine cache implementations depending on redis is available or not (use sqlite then as fallback).

Reply:
If the SQLite solution is kept, here some links that might help to avoid some already seen problems:
* https://github.com/piskvorky/sqlitedict
* https://stackoverflow.com/questions/47233562/key-value-store-in-python-for-possibly-100-gb-of-data-without-client-server

(there is [lmdb](https://lmdb.readthedocs.io/en/release/) too, but there is no wheels for the ARMv7 architecture)